zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,324 @@
+package org.apache.zookeeper.backup;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FilenameFilter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.DirectoryNotEmptyException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.apache.zookeeper.server.persistence.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Implementation for NFS-based backup storage provider
+ */
+public class NFSBackupStorage implements BackupStorageProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(NFSBackupStorage.class);
+
+  public static final String TMP_FILE_PREFIX = "TMP_";
+  private static final String PATH_SEPARATOR = File.separator;
+  private final BackupConfig _backupConfig;
+  private final String _fileRootPath;
+
+  /**
+   * Constructor using BackupConfig to get backup storage info
+   * @param backupConfig The information and settings about backup storage, to be set as a part of ZooKeeper server config
+   */
+  public NFSBackupStorage(BackupConfig backupConfig) {
+    _backupConfig = backupConfig;
+    _fileRootPath =
+        String.join(PATH_SEPARATOR, _backupConfig.getMountPath(), _backupConfig.getNamespace());
+  }
+
+  @Override
+  public BackupFileInfo getBackupFileInfo(File file) throws IOException {
+    String backupFilePath = constructBackupFilePath(file.getName());
+    File backupFile = new File(backupFilePath);
+
+    if (!backupFile.exists()) {
+      throw new BackupException(
+          "Backup file " + file.getName() + " does not exist, could not get file info.");
+    }
+
+    BasicFileAttributes fileAttributes =
+        Files.readAttributes(Paths.get(backupFilePath), BasicFileAttributes.class);
+    return new BackupFileInfo(backupFile, fileAttributes.lastModifiedTime().toMillis(),
```

Comment:
I remember as we discussed, better to have something like "getAndCreate". Not sure whether this is the right place. But this method only returns the info without creating new files.

Reply:
Not sure if this is what you meant. I modified the method and now if the file does not exist, we still return a BackupFileInfo instance with lastModifiedTime and fileSize to be -1 indicating that this file does not exist.
Are you saying we want to create a new file after we find out that it does not exist when calling getFileInfo()? If so, let's discuss offline because I think the only place we should create new files is when we copy files to backup storage. :)